### ISSUE TITLE
Add overloads for validation problem results that accept IReadOnlyDictionary<string, string[]>

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/aspnetcore#41899]
Sometimes all you have is a `IReadOnlyDictionary` and you don't want to have to downcast it or wrap it to pass it to the helper methods.

```diff
namespace Microsoft.AspNetCore.Http;

public static class Results
{
+    public static IResult ValidationProblem(IReadOnlyDictionary<string, string[]> errors,
+        string? detail = null,
+        string? instance = null,
+        int? statusCode = null,
+        string? title = null,
+        string? type = null,
+        IReadOnlyDictionary<string, object?>? extensions = null);

+    public static IResult Problem(
+        string? detail = null,
+        string? instance = null,
+        int? statusCode = null,
+        string? title = null,
+        string? type = null,
+        IReadOnlyDictionary<string, object?>? extensions = null);
}

public static class TypedResults
{
+    public static ValidationProblem ValidationProblem(IReadOnlyDictionary<string, string[]> errors,
+        string? detail = null,
+        string? instance = null,
+        string? title = null,
+        string? type = null,
+        IReadOnlyDictionary<string, object?>? extensions = null);

+    public static ProblemHttpResult Problem(
+        string? detail = null,
+        string? instance = null,
+        string? title = null,
+        string? type = null,
+        IReadOnlyDictionary<string, object?>? extensions = null);
}
```

**Comment by captainsafia** [comment]
Triage: Marking this as ready for API review. Once it is approved for API review, this would be a good candidate for `help wanted`.

**Comment by halter73** [comment]
API Review Notes:

- Is this a breaking change? [Yes.](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWoGYEBMcAwnAN5ZxUL5TJwCyAFAJIBKApgIbADyAdgBsAngBEAlgGMALuID2/LgCdhAHiQwANAkQwA2gF0AfHGBTZC5cICU5StQC+9qs5oJ6zFhJnzFK9braGoYmZj6WKrYUmNRwTjHUrnjucAAqHADO0kzWrtGxscz8HADucN4WfmoaQbohOdYYCVTxDkA) It's ambiguous which overload to pick if you pass in `Dictionary<string, string[]>`. 
- Can we make it `IEnumerable<KeyValuePair<string, string[]>>` instead? [Yes.](https://sharplab.io/#v2:CYLg1APgAgDABFAjAFgNwFgBQWoGYEBMcAwnAN5ZxUL5TJwCyAFErgDwDSApgJ4BqAQwA2AVy4AFAQEsATmyQwANAkQwA2gF0AfFrjApAYwAuUgPYA7ATJ4BKStQqZqcAL72q7mgnrMAkgBFDEwsrHnlVZQVNXX1jM0trOycHTzdkj3SvOjgAFS4AZyMmJOdHZ2dmJl8AJS4BYAB5cyEeQLiQ63ClFXVtGzhzLgB3ODbghLCFSNVo4psMTLSXIA=)
- Do we want to change both the `errors` and `extensions` parameters? Might as well. It will accept more arguments, and we're not modifying either dictionary.

API Approved!

```diff
namespace Microsoft.AspNetCore.Http;

public static class Results
{
+    public static IResult ValidationProblem(IEnumerable<KeyValuePair<string, string[]>> errors,
+        string? detail = null,
+        string? instance = null,
+        int? statusCode = null,
+        string? title = null,
+        string? type = null,
+        IEnumerable<KeyValuePair<string, object?>>? extensions = null);

+    public static IResult Problem(
+        string? detail = null,
+        string? instance = null,
+        int? statusCode = null,
+        string? title = null,
+        string? type = null,
+        IEnumerable<KeyValuePair<string, object?>>? extensions = null);
}

public static class TypedResults
{
+    public static ValidationProblem ValidationProblem(IEnumerable<KeyValuePair<string, string[]>> errors,
+        string? detail = null,
+        string? instance = null,
+        string? title = null,
+        string? type = null,
+        IEnumerable<KeyValuePair<string, object?>>? extensions = null);

+    public static ProblemHttpResult Problem(
+        string? detail = null,
+        string? instance = null,
+        string? title = null,
+        string? type = null,
+        IEnumerable<KeyValuePair<string, object?>>? extensions = null);
}
```

**Comment by joegoldman2** [comment]
I'm trying to implement this proposal but I have two questions:
- For all new overloads I'm getting the following warning: `RS0026 Symbol 'ValidationProblem' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'. See 'https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md'`. Is it ok?
- I think the parameter `int? statusCode = null` is missing in the 2 new overloads for `TypedResults`.

**Comment by joegoldman2** [comment]
I'm trying to implement this proposal but I have two questions:
- For all new overloads I'm getting the following warning: `RS0026 Symbol 'ValidationProblem' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'. See 'https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md'`. Is it ok?
- I think the parameter `int? statusCode = null` is missing for the new `Problem` overload in `TypedResults`.

**Comment by danirzv** [comment]
hello,
I've done changes mentioned but in running tests find some errors which i have no idea how should i change tests as blow:
```
  Microsoft.AspNetCore.Http.Results failed with 8 error(s) (1.1s)
    D:\Git\aspnetcore\src\Http\Http.Results\src\TypedResults.cs(758,37): error RS0016: Symbol 'Problem' is not part of the declared API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\TypedResults.cs(813,37): error RS0016: Symbol 'ValidationProblem' is not part of the declared API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\Results.cs(715,27): error RS0016: Symbol 'Problem' is not part of the declared API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\Results.cs(744,27): error RS0016: Symbol 'ValidationProblem' is not part of the declared API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\PublicAPI.Shipped.txt(271,1): error RS0017: Symbol 'static Microsoft.AspNetCore.Http.Results.Problem(string? detail = null, string? instance = null, int? statusCode = null, string? title = null, string? type = null, System.Collections.Generic.IEnumerable<KeyValuePair<string, object?>>? extensions = null) -> Microsoft.AspNetCore.Http.IResult!' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\PublicAPI.Shipped.txt(287,1): error RS0017: Symbol 'static Microsoft.AspNetCore.Http.Results.ValidationProblem(System.Collections.Generic.IDictionary<string!, string![]!>! errors, string? detail = null, string? instance = null, int? statusCode = null, string? title = null, string? type = null, System.Collections.Generic.IEnumerable<KeyValuePair<string, object?>>? extensions = null) -> Microsoft.AspNetCore.Http.IResult!' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\PublicAPI.Shipped.txt(331,1): error RS0017: Symbol 'static Microsoft.AspNetCore.Http.TypedResults.Problem(string? detail = null, string? instance = null, int? statusCode = null, string? title = null, string? type = null, System.Collections.Generic.IDictionary<string!, object?>? extensions = null) -> Microsoft.AspNetCore.Http.HttpResults.ProblemHttpResult!' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
    D:\Git\aspnetcore\src\Http\Http.Results\src\PublicAPI.Shipped.txt(347,1): error RS0017: Symbol 'static Microsoft.AspNetCore.Http.TypedResults.ValidationProblem(System.Collections.Generic.IDictionary<string!, string![]!>! errors, string? detail = null, string? instance = null, string? title = null, string? type = null, System.Collections.Generic.IDictionary<string!, object?>? extensions = null) -> Microsoft.AspNetCore.Http.HttpResults.ValidationProblem!' is part of the declared API, but is either not public or could not be found (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md)
```

**Comment by danirzv** [comment]
Can anyone assist me with this problem?

**Comment by joegoldman2** [comment]
@danirzv the goal is not to replace existing overloads but add new ones. Let's wait for them to reply to https://github.com/dotnet/aspnetcore/issues/41899#issuecomment-2065668446 first.

**Comment by danirzv** [comment]
> @danirzv the goal is not to replace existing overloads but add new ones. Let's wait for them to reply to https://github.com/dotnet/aspnetcore/issues/41899#issuecomment-2065668446 first.

As @halter73 mentioned it's a breaking change I thought we want to replace them, but you are right, i have to wait for answer

**Comment by joegoldman2** [comment]
It's a breaking change with overloads that accept a `IReadOnlyDictionary<string, string[]>` because of the ambiguity with the existing overloads. There is no ambiguity with the approved API (`IEnumerable<KeyValuePair<string, string[]>>`).

**Comment by joegoldman2** [comment]
@halter73 can you please take a look at this comment please? https://github.com/dotnet/aspnetcore/issues/41899#issuecomment-2065668446

**Comment by halter73** [comment]
> * For all new overloads I'm getting the following warning: `RS0026 Symbol 'ValidationProblem' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'. See 'https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md'`. Is it ok?

It's not okay. While we were not explicit about this in our API approval, you should always remove the default values from the old overload to make them non-optional when adding a new overload (which should always be longer or more permissive). Only the new public overload should have optional parameters with default values. It would introduce a breaking change to do otherwise.

The API analyzer will complain about modifying public API, but as noted in the [doc linked to by the error message](https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md), you can "copy the entry for your change from PublicAPI.Shipped.txt and then put that entry in PublicAPI.Unshipped.txt with `*REMOVED*` prefix."

> * I think the parameter `int? statusCode = null` is missing for the new `Problem` overload in `TypedResults`. It exists for the existing overload that accepts a `IDictionary<string, object?>?`.

Good catch. I updated the API approval to include `int? statusCode = null` for this overload.

> * think a new construtor overload that accepts a `IEnumerable<KeyValuePair<string, string[]>>` for `HttpValidationProblemDetails` is welcome

I also think this would be a good idea for consistency, but it would require a separate API approval. Please file a [new API proposal issue](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=) suggesting it. It doesn't have to be too detailed. And feel free to link to this issue and include consistency as the motivation.

**Comment by joegoldman2** [comment]
Thank you for your reply @halter73.

I still get the same warning `RS0026 Symbol 'ValidationProblem' violates the backcompat requirement: 'Do not add multiple overloads with optional parameters'` with the following entries in PublicAPI.Unshipped.txt.

```
static Microsoft.AspNetCore.Http.TypedResults.Problem(string? detail = null, string? instance = null, int? statusCode = null, string? title = null, string? type = null, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string!, object?>>? extensions = null) -> Microsoft.AspNetCore.Http.HttpResults.ProblemHttpResult!
*REMOVED*static Microsoft.AspNetCore.Http.TypedResults.Problem(string? detail = null, string? instance = null, int? statusCode = null, string? title = null, string? type = null, System.Collections.Generic.IDictionary<string!, object?>? extensions = null) -> Microsoft.AspNetCore.Http.HttpResults.ProblemHttpResult!
static Microsoft.AspNetCore.Http.TypedResults.Problem(string? detail, string? instance, int? statusCode, string? title, string? type, System.Collections.Generic.IDictionary<string!, object?>? extensions) -> Microsoft.AspNetCore.Http.HttpResults.ProblemHttpResult!
```

The first one adds the overload that takes a  `IEnumerable<KeyValuePair<string, string[]>>`. The second one removes the current overload with `IDictionary<string, object?>?` and optional params. The third one adds the overload with `IDictionary<string, object?>?` without optional params.

**Comment by halter73** [comment]
That looks right to me, but I haven't tried it. Feel free to submit a PR with the error and we can take a look.

**Comment by halter73** [comment]
> * I think a new construtor overload that accepts a `IEnumerable<KeyValuePair<string, string[]>>` for `HttpValidationProblemDetails` is welcome in addition to the existing one which accepts a `IDictionary<string, string[]>`. This will make it easier to create a new instance from a `IEnumerable<KeyValuePair<string, string[]>>` and will also ensure a degree of consistency.
>   https://github.com/dotnet/aspnetcore/blob/c7aae8ff34dce81132d0fb3a976349dcc01ff903/src/Http/Http.Abstractions/src/ProblemDetails/HttpValidationProblemDetails.cs#L26

Let's update this API proposal to include a new `HttpValidationProblemDetails` constructor. The alternative requires making Microsoft.AspNetCore.Http.Abstractions [`<InternalsVisibleTo Include="Microsoft.AspNetCore.Http.Results" />`](https://github.com/dotnet/aspnetcore/pull/56368/files#diff-aa687fb4795f3f50c0239c5b9b0b31c3e8271ab05dad9826f971aa220b00b227R40) which we'd rather avoid. It makes sense for consistency too as you noted.

```diff
// Microsoft.AspNetCore.Http.Abstractions.dll

namespace Microsoft.AspNetCore.Http;

public class HttpValidationProblemDetails : ProblemDetails
{
    public HttpValidationProblemDetails()
        : this(new Dictionary<string, string[]>(StringComparer.Ordinal));

    public HttpValidationProblemDetails(IDictionary<string, string[]> errors)
        : this(new Dictionary<string, string[]>(errors ?? throw new ArgumentNullException(nameof(errors)), StringComparer.Ordinal));

+    public HttpValidationProblemDetails(IEnumerable<KeyValuePair<string, string[]>> errors)
+        : this(new Dictionary<string, string[]>(errors ?? throw new ArgumentNullException(nameof(errors)), StringComparer.Ordinal));
}
```

**Comment by joegoldman2** [comment]
@halter73 thank you for reopening this one. I already opened #56370 about the new constructor for `HttpValidationProblemDetails`.

**Comment by halter73** [comment]
Thanks. I missed that. Let's use #56370 to track the update. I put back the `api-approved` label to indicate that the API in https://github.com/dotnet/aspnetcore/issues/41899#issuecomment-1476727981 is still approved.
